proposals/webidl-bindings/idl.md

```diff
@@ -0,0 +1,739 @@
+Signature => 'all' TypeVar .. TypeVar '.' NominalType '=' Type
+```
+
+```
+W-Signature ::= '(' 'type' W-NominalType W-Type ')'
+W-Signature ::= '(' 'type' '(' 'all' '(' TypeVar .. TypeVar ')' W-NominalType W-Type ')' ')'
+```
+
+For example, the standard `Option/1` type can be defined as the _Signature_:
+
+```
+( type (all (t) <Option t> [cases none (some t)]))
+```
+
+#### Tuple Type {#tuple-type}
+
+A TupleType is a linear sequence of types. It denotes the cross product of its
+component types:
+
+```
+TupleType => '(' Type ..  Type ')'
+```
+
+```
+W-TupleType ::= '(' W-Type .. W-Type ')'
+```
+
+There are two special cases of _TupleType_ that warrant further discussion: the
+unary tuple and the zero-ary tuple.
+
+A unary tuple type, such as
+
+```
+(integer)
+```
+
+is _not_ equivalent to its element type.
+
+The zero tuple type:
+
+```
+()
+```
+is used to designate void.
+
+>NOTE: Due to the ubiquity of tuples, the W-Expression form of a tuple type has
+>a special status: it is not marked with a special keyword as the first element
+>of its written form.
+
+#### Variant Type {#type-union}
+
+
+A variant type -- sometimes called a discriminated union -- is a collection of
+alternate forms of type; each of which is labeled by a discriminator.
+
+```
+VariantType => Name Type '|' .. '|' Name Type
+```
+
+If the Type is the empty tuple then it may be omitted.
+
+```
+W-VariantType ::= '[' 'cases' W-Variant .. W-Variant ']'
+W-Variant ::= W-Identifier | '(' W-Identifier W-Type ')'
+```
+
+<div class="example">
+The legal values for standard Boolean type may be defined as a VariantType:
+
+```
+[cases True False]
+```
+</div>
+
+If the Type is a TypInterface then the variant counts as the description of a
+record type; if the Type is a TupleType then the variant counts as a form of
+labeled tuple.
+
+#### Nominal Type {#nominal-type}
+
+A nominal type expression is used to identify types by name:
+
+```
+NominalType => Name
+NominalType => Name '<' Type ..  Type '>'
+```
+
+```
+W-NominalType ::= W-Identifier
+W-NominalType ::= '<' W-Identifier W-Type .. W-Type '>'
+```
+
+Any _Name_ appearing in a _NominalType_ must either be a standard name or it must
+have appeared in scope; either as a _Signature_ or as a bound _TypeVar_.
+
+## Standard Types {#standard-types}
+
+Standard types are nominal types whose scope encompasses all type expressions;
+i.e., it is as though any given type expression were wrapped in an all
+quantification that mentions all the standard types.
+
+### Numeric Types{#numeric-types}
+
+This includes `Integer` and `Float`.
+
+Note that we specifically do not elaborate the numeric types -- for example into
+signed vs unsigned, or 8bit vs 64bit. Such elaboration can be accounted for by
+appropriate operators.
+
+### Boolean Type {#boolean-type}
+
+The `Boolean` type is standard, but is defined as though by an TypeDefinition
+signature:
+
+<div class="example">
+```
+(type Boolean [cases true false])
```

Comment:
This may become problematic if you ever want to add subtyping to the IDL, since it would force Bool to be compatible with any variant super and subtype.